crates/oxc_linter/src/rules/eslint/no_control_regex.rs

```diff
@@ -0,0 +1,351 @@
+                            _ => {
+                                continue;
+                            }
+                        };
+
+                        // 1. Unicode control pattern is missing a curly brace
+                        //    and is therefore invalid. (note: we may want to
+                        //    report this?)
+                        // 2. Unicode flag is missing, which is needed for
+                        //    interpreting \u{`nn`} as a unicode character
+                        if !has_unicode_flag || !numeric_part.ends_with('}') {
+                            continue;
+                        }
+
+                        numeric_part = &numeric_part[1..numeric_part.len() - 1];
+                    }
+
+                    match u32::from_str_radix(numeric_part, 16) {
+                        Err(_) => continue,
+                        Ok(as_num) if as_num > 0x1f => continue,
+                        Ok(_) => { /* noop */ }
+                    }
+                }
+
+                violations.push(ctl);
+            }
+
+            if !violations.is_empty() {
+                let violations = violations.join(", ");
+                context.diagnostic(NoControlRegexDiagnostic(violations.into(), span));
+            }
+        }
+    }
+}
+
+fn extract_flags<'a>(args: &'a oxc_allocator::Vec<'a, Argument<'a>>) -> Option<RegExpFlags> {
+    if args.len() > 1 &&
+    let Argument::Expression(Expression::StringLiteral(flag_arg)) = &args[1] {
+        // TODO: how should we handle invalid flags?
+        let mut flags = RegExpFlags::empty();
+        for ch in flag_arg.value.chars() {
+            let flag = RegExpFlags::try_from(ch).ok()?;
+            // TODO: should we check for duplicates?
+            flags |= flag;
+        }
+        Some(flags)
+    } else {
+        None
+    }
+}
+
+/// Returns the regex pattern inside a node, if it's applicable.
+///
+/// e.g.:
+/// * /foo/ -> "foo"
+/// * new RegExp("foo") -> foo
+///
+/// note: [`RegExpFlags`] and [`Span`]s are both tiny and cloneable.
+fn regex_pattern<'a>(node: &AstNode<'a>) -> Option<(&'a Atom, Option<RegExpFlags>, Span)> {
+    let kind = node.kind();
+    match kind {
+        // regex literal
+        AstKind::RegExpLiteral(reg) => Some((&reg.regex.pattern, Some(reg.regex.flags), reg.span)),
+
+        // FIXME: we need a more graceful way to handle NewExpr/CallExprs
+
+        // new RegExp()
+        AstKind::NewExpression(expr) => {
+            // constructor is RegExp,
+            if let Expression::Identifier(ident) = &expr.callee &&
+            ident.name == "RegExp" &&
+            // which is provided at least 1 parameter,
+            expr.arguments.len() > 0 &&
+            // where the first one is a string literal
+            // note: improvements required for strings used via identifier
+            // references
+            let Argument::Expression(Expression::StringLiteral(pattern)) = &expr.arguments[0]
```

Comment:
For repeated code here, you can add a trait or a helper function.

Reply:
> For repeated code here, you can add a trait or a helper function.

I don't think I'm _quite_ practiced enough with lifetimes to do this without some kind of copy